File: killrvideo-commons/src/main/java/com/killrvideo/utils/GrpcMappingUtils.java
package com.killrvideo.utils;

import java.time.Instant;
import java.util.Date;
import java.util.UUID;

import com.datastax.oss.driver.api.core.uuid.Uuids;
import com.google.protobuf.Timestamp;

import killrvideo.common.CommonTypes.TimeUuid;
import killrvideo.common.CommonTypes.Uuid;

/**
 * Mapping.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
public class GrpcMappingUtils {
    
    /** Hiding private constructor. */
    private GrpcMappingUtils() {}
    
    /**
     * Conversions.
     */
    public static TimeUuid uuidToTimeUuid(UUID uuid) {
        return TimeUuid.newBuilder().setValue(uuid.toString()).build();
    }

    public static UUID fromTimeUuid(TimeUuid timeUuid) {
        return UUID.fromString(timeUuid.getValue());
    }
    
    public static Uuid uuidToUuid(UUID uuid) {
        return Uuid.newBuilder().setValue(uuid.toString()).build();
    }

    public static UUID fromUuid(Uuid uuid) {
        return UUID.fromString(uuid.getValue());
    }

    public static Uuid randomUuid() {
        return uuidToUuid(UUID.randomUUID());
    }

    public static TimeUuid randomTimeUuid() {
        return uuidToTimeUuid(Uuids.timeBased());
    }
    
    public static Date timestampToDate(Timestamp protoTimestamp) {
        return Date.from(timestampToInstant(protoTimestamp));
    }

    public static Timestamp dateToTimestamp(Date date) {
        return instantToTimeStamp(date.toInstant());
    }

    public static Instant timestampToInstant(Timestamp protoTimeStamp) {
        return Instant.ofEpochSecond(
                protoTimeStamp.getSeconds(),
                protoTimeStamp.getNanos() ) ;
    }

    public static Timestamp instantToTimeStamp(Instant instant) {
        return Timestamp.newBuilder().setSeconds(instant.getEpochSecond()).setNanos(instant.getNano()).build();
    }
}

File: killrvideo-commons/src/main/kotlin/com/killrvideo/utils/GrpcMappingUtils.kt
package com.killrvideo.utils

import com.datastax.oss.driver.api.core.uuid.Uuids
import com.google.protobuf.Timestamp
import killrvideo.common.CommonTypes
import killrvideo.common.CommonTypes.TimeUuid
import java.time.Instant
import java.util.*

/**
 * Mapping.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
object GrpcMappingUtils {
    /**
     * Conversions.
     */
    fun uuidToTimeUuid(uuid: UUID): TimeUuid =
        TimeUuid.newBuilder().setValue(uuid.toString()).build()

    fun fromTimeUuid(timeUuid: TimeUuid): UUID =
        UUID.fromString(timeUuid.value)

    fun uuidToUuid(uuid: UUID): CommonTypes.Uuid =
        CommonTypes.Uuid.newBuilder().setValue(uuid.toString()).build()

    fun fromUuid(uuid: CommonTypes.Uuid): UUID =
        UUID.fromString(uuid.value)

    fun randomUuid(): CommonTypes.Uuid =
        uuidToUuid(UUID.randomUUID())

    fun randomTimeUuid(): TimeUuid =
        uuidToTimeUuid(Uuids.timeBased())

    fun timestampToDate(protoTimestamp: Timestamp): Date =
        Date.from(timestampToInstant(protoTimestamp))

    @JvmStatic
    fun dateToTimestamp(date: Date): Timestamp =
        instantToTimeStamp(date.toInstant())

    @JvmStatic
    fun timestampToInstant(protoTimeStamp: Timestamp): Instant =
        Instant.ofEpochSecond(
            protoTimeStamp.seconds,
            protoTimeStamp.nanos.toLong()
        )

    @JvmStatic
    fun instantToTimeStamp(instant: Instant): Timestamp =
        Timestamp.newBuilder().setSeconds(instant.epochSecond).setNanos(instant.nano).build()
}
